plugins/tech-insights-backend/src/plugin/plugin.ts

```diff
@@ -0,0 +1,128 @@
+ */
+
+import { loggerToWinstonLogger } from '@backstage/backend-common';
+import {
+  coreServices,
+  createBackendPlugin,
+} from '@backstage/backend-plugin-api';
+import { CheckResult } from '@backstage/plugin-tech-insights-common';
+import {
+  FactCheckerFactory,
+  FactRetriever,
+  FactRetrieverRegistration,
+  TechInsightCheck,
+  techInsightsFactCheckerFactoryExtensionPoint,
+  techInsightsFactRetrieversExtensionPoint,
+} from '@backstage/plugin-tech-insights-node';
+import {
+  buildTechInsightsContext,
+  createRouter,
+  entityMetadataFactRetriever,
+  entityOwnershipFactRetriever,
+  techdocsFactRetriever,
+} from '../service';
+import { createFactRetrieverRegistrationFromConfig } from './config';
+
+/**
+ * The tech-insights backend plugin.
+ *
+ * @public
+ */
+export const techInsightsPlugin = createBackendPlugin({
+  pluginId: 'tech-insights',
+  register(env) {
+    let factCheckerFactory:
+      | FactCheckerFactory<TechInsightCheck, CheckResult>
+      | undefined = undefined;
+    env.registerExtensionPoint(techInsightsFactCheckerFactoryExtensionPoint, {
+      setFactCheckerFactory<
+        CheckType extends TechInsightCheck,
+        CheckResultType extends CheckResult,
+      >(factory: FactCheckerFactory<CheckType, CheckResultType>): void {
+        factCheckerFactory = factory;
+      },
+    });
+
+    // initialized with built-in fact retrievers
+    // only added as registration if there is config for them
+    const addedFactRetrievers: Record<string, FactRetriever> = {
+      entityMetadataFactRetriever,
+      entityOwnershipFactRetriever,
+      techdocsFactRetriever,
+    };
+    env.registerExtensionPoint(techInsightsFactRetrieversExtensionPoint, {
+      addFactRetrievers(factRetrievers: Record<string, FactRetriever>): void {
+        Object.entries(factRetrievers).forEach(([key, value]) => {
+          addedFactRetrievers[key] = value;
+        });
+      },
+    });
+
+    env.registerInit({
+      deps: {
+        config: coreServices.rootConfig,
+        database: coreServices.database,
+        discovery: coreServices.discovery,
+        httpRouter: coreServices.httpRouter,
+        logger: coreServices.logger,
+        scheduler: coreServices.scheduler,
+        tokenManager: coreServices.tokenManager,
+      },
+      async init({
+        config,
+        database,
+        discovery,
+        httpRouter,
+        logger,
+        scheduler,
+        tokenManager,
+      }) {
+        const winstonLogger = loggerToWinstonLogger(logger);
+        const factRetrievers: FactRetrieverRegistration[] = Object.entries(
+          addedFactRetrievers,
+        )
+          .map(([name, factRetriever]) =>
+            createFactRetrieverRegistrationFromConfig(
+              config,
+              name,
+              factRetriever,
+            ),
+          )
+          .filter(registration => registration) as FactRetrieverRegistration[];
+
+        const context = await buildTechInsightsContext({
```

Comment:
This construction looks to omit the possibility to inject alternative optional storage solutions. Are those omitted on purpose or forgotten?

Reply:
I'm aware that I might not cover all use cases with this PR, yet.

I assume you refer to 
```ts
persistenceContext?: PersistenceContext;
```

Additionally,
```ts
factRetrieverRegistry?: FactRetrieverRegistry
```
is not yet covered, too.

I think we could add support at follow-up PRs. However, I can look into what we would need. Maybe, it is good enough to provide more extension points for these. 👀

Reply:
Yup indeed. We are using both of those to use a bit enhanced data storage option and a more multifunctional FactRetrieverRegistry. Could probably be fine as injectables/extension points, they are conceptually big but kinda important for our use case to be able to replace them with alternative implementations

Reply:
I will need to move these interfaces to the `-node` package to be able to add extension points. I will add a commit adding support for both in a bit.

Reply:
In the end, it was rather simple to add the additional extension points. I will add a separate changeset for the moved types.